src/frontend/src/binder/expr/mod.rs

```diff
@@ -561,6 +562,18 @@ impl Binder {
         let lhs = self.bind_expr_inner(expr)?;
         lhs.cast_explicit(data_type).map_err(Into::into)
     }
+
+    pub fn bind_collate(&mut self, expr: Expr, collation: ObjectName) -> Result<ExprImpl> {
+        if !["C", "POSIX"].contains(&collation.real_value().as_str()) {
+            return Err(ErrorCode::NotImplemented(
+                "Collate collation other than `C` or `POSIX` is not implemented".into(),
+                None.into(),
+            )
+            .into());
+        }
+
+        self.bind_expr_inner(expr)
```

Comment:
If we bind the inner expr first, we can `if bound_inner.return_type() != DataType::Varchar`. The `create_table` handler has access to `data_type` as well. Not required, but seems easy to achieve.

Btw found an interesting case in PostgreSQL:
```
test=# select '1' collate "C" > 2;
 ?column? 
----------
 f
(1 row)
```
It considers `'1'` as collatable initially, then converts it into an interger for `>`. Our implementation behaves the same way.